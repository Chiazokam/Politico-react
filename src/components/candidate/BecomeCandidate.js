import React, {Component} from 'react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { Container, Button, ToastMessage } from '../global';
import { getOffices, getParties, indicateInterest } from '../../actions';
import { getUser } from '../../utils';
import '../../styles/create-party/create-party.scss';

class BecomeCandidate extends Component {
  constructor(props) {
    super(props);
    this.state = {
      office: '',
      party: '',
      userId: ''
    };
  }

  componentDidMount() {
    this.getOffices();
    this.getParties();
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const user = getUser()
    const userId = String(user.id);
    const candidate = {
      party: this.state.party,
      office: this.state.office,
      userId,
    };
    this.props.dispatch(indicateInterest(candidate));
  }

  getOffices = () => {
    this.props.dispatch(getOffices());
  }
  
  getParties = () => {
    this.props.dispatch(getParties());
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value});
  }

  render() {
    const { getOffices, getParties, submit, redirect, errors } = this.props;
    if (redirect) {
      toast(<ToastMessage message={'Successfully declared interest'} />, {
        type: 'success',
        closeButton: false,
        hideProgressBar: true,
        autoClose: 0,
      });
    }
    
    return (
      <div className="interest">
        <Container>
          <p className="form-text">Declare Interest</p>
            <form onSubmit={this.handleSubmit}>
            <div className="error">{errors.message}</div>
              <div className="form-field">
              <div className="error">{errors.office}</div>
                  <select id="office-output" onChange={this.handleChange} name='office'>
                    {getOffices.length === 0 ? <option>No offices currently</option> : <option>Select Office</option> }
                    { getOffices.map(office =>
                      <option key={office.id} value={office.id}>{office.name}</option>
                    )}
                  </select> 
              </div>

              <div className="form-field">
              <div className="error">{errors.party}</div>
                  <select id="party-output" onChange={this.handleChange}  name='party'>
                    {getParties.length === 0 ? <option>No Parties currently</option> : <option>Select Party</option> }
                    { getParties.map(party =>
                      <option key={party.id} value={party.id}>{party.name}</option>
                    )}
                  </select>
              </div>
              <Button
                type='submit'
                className='btn btn-colored btn-signin btn-dark'>
                { submit && <FontAwesomeIcon 
                  icon={ faSpinner }
                  spin
                /> }
                &nbsp; Submit
              </Button>
            </form>
        </Container>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  getOffices: state.office.getOffices,
  getParties: state.party.getParties,
  submit: state.candidate.submit,
  redirect: state.candidate.redirect,
  errors: state.candidate.errors
});

export default connect(mapStateToProps)(BecomeCandidate);
