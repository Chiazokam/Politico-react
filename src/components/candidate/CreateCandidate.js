import React, {Component} from 'react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { Container, Button, ToastMessage } from '../global';
import { getCandidatesRequest, createCandidate, clearErrors } from '../../actions';
import '../../styles/create-party/create-party.scss';

class CreateCandidate extends Component {
  constructor(props) {
    super(props);
    this.state = {
      office: '',
      party: '',
      user: '',
      error: '',
      success: ''
    };
  }

  componentDidMount() {
    this.getCandidates();
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const candidate = {
      party: this.state.party,
      office: this.state.office
    }
    const userid = this.state.user;

    if (candidate.party === '' || candidate.office === '' || userid === '') {
      this.setState({ error: 'Field cannot be empty' })
    } else {
      this.props.dispatch(createCandidate(candidate, userid));
    }
  }

  getCandidates = () => {
    this.props.dispatch(getCandidatesRequest());
  }

  handleChange = (e) => {
    const { candidates, dispatch } = this.props;
    this.setState({ 
      [e.target.name]: e.target.value,
      error: ''  
    });
    dispatch(clearErrors());
    const userid = parseInt(e.target.value)
    var data = candidates.filter(candidate => {
      return candidate.userid === userid
    })
    const party = String(data[0].party);
    const office = String(data[0].office);
    this.setState({ 
      party,
      office,
    });
  }

  render() {
    const { candidates, errors, submitCandidate, redirect } = this.props;
    if (redirect) {
      toast(<ToastMessage message={'Successfully created a candidate'} />, {
        type: 'success',
        closeButton: false,
        hideProgressBar: true,
        autoClose: 0,
      });
    }

    return (
      <div className="interest">
        <Container>
          <p className="form-text">Create Candidate</p>
            <form onSubmit={this.handleSubmit}>
            <div className="error">{errors.message || this.state.error} </div>
              <div className="form-field">
                  <select onChange={this.handleChange} name='user'>
                    {candidates.length === 0 ? <option>No interested candidates currently</option> : <option>Select Candidate</option> }
                    { candidates.map(candidate =>
                      <option
                        key={candidate.id}
                        value={candidate.userid}
                      >
                        {candidate.firstname} {candidate.lastname}
                      </option>
                    )}
                  </select> 
              </div>

              <Button
                type='submit'
                className='btn btn-colored btn-signin btn-dark'>
                { submitCandidate && <FontAwesomeIcon 
                  icon={ faSpinner }
                  spin
                /> }
                &nbsp; Submit
              </Button>
            </form>
        </Container>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  candidates: state.candidate.candidates,
  errors: state.candidate.candidateErrors,
  submitCandidate: state.candidate.submitCandidate,
  redirect: state.candidate.candidateRedirect
});

export default connect(mapStateToProps)(CreateCandidate);
