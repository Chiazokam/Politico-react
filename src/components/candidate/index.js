import { BecomeCandidate } from './BecomeCandidate';
import { CreateCandidate } from './CreateCandidate'

export { BecomeCandidate, CreateCandidate };
