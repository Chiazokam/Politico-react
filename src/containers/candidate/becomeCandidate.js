import React from 'react';
import { Header, Footer } from '../../components/global';
import { BecomeCandidate } from '../../components/candidate';

const Layout = () => {
  const navItems = ['logout'];
  return (
    <React.Fragment>
      <Header navItems={navItems}/>
      <BecomeCandidate />
      <Footer />
    </React.Fragment>
    )
}

export default Layout;
