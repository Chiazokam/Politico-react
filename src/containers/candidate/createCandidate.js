import React from 'react';
import { Header, Footer } from '../../components/global';
import { CreateCandidate } from '../../components/candidate';

const Layout = () => {
  const navItems = ['logout'];
  return (
    <React.Fragment>
      <Header navItems={navItems}/>
      <CreateCandidate />
      <Footer />
    </React.Fragment>
    )
}

export default Layout;
