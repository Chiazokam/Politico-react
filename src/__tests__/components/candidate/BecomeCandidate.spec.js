import React from 'react';
import "@babel/polyfill";
import { shallow } from 'enzyme';
import { BecomeCandidateUnit } from '../../../components/candidate/BecomeCandidate';

describe('<Become Candidate />', () => {
  it('renders correctly', () => {
    const props = {
      errors: {
        message: 'Wrong input',
        office: 'Chief Judge',
        party: 'New Party',
      },
      getOffices: [],
      getParties: [],
      offices: jest.fn(),
      parties: jest.fn()
    };
    const tree = shallow(<BecomeCandidateUnit {...props} />);
    expect(tree).toMatchSnapshot(); 
  });
});
