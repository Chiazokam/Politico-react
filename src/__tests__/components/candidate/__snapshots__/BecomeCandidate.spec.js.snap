// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Become Candidate /> renders correctly 1`] = `ShallowWrapper {}`;
